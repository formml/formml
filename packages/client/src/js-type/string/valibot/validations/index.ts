import bool from './bool.js'
import datetime from './datetime.js'
import decimal from './decimal.js'
import num from './num.js'

export { bool, datetime, decimal, num }
