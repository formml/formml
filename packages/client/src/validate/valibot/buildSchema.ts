import { Field } from '@formml/dsl'

export default function buildSchema(formmlSchema: Field) {
  return formmlSchema
}
