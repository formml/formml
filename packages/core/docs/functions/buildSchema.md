[**@formml/core**](../README.md) • **Docs**

---

[@formml/core](../globals.md) / buildSchema

# Function: buildSchema()

## buildSchema(formmlSchema, preprocess)

> **buildSchema**\<`T`\>(`formmlSchema`, `preprocess`?): `v.GenericSchema`\<[`PrimitiveTypeMapping`](../namespaces/JSType/type-aliases/PrimitiveTypeMapping.md)\[`T`\]\>

### Type Parameters

• **T** _extends_ `PRIMITIVE`

### Parameters

• **formmlSchema**: `Field`\<`T`\>

• **preprocess?**: `Record`\<`PRIMITIVE`, `GenericSchema`\<`unknown`, `unknown`, `BaseIssue`\<`unknown`\>\>\>

### Returns

`v.GenericSchema`\<[`PrimitiveTypeMapping`](../namespaces/JSType/type-aliases/PrimitiveTypeMapping.md)\[`T`\]\>

### Defined in

[validator/buildSchema.ts:45](https://github.com/formml/formml/blob/fed46848d8032d8aeab7f7fad75fbc02dc65656a/packages/core/src/validator/buildSchema.ts#L45)

## buildSchema(formmlSchema, preprocess)

> **buildSchema**(`formmlSchema`, `preprocess`?): `v.StrictObjectSchema`\<`Record`\<`string`, `v.GenericSchema`\>, `string`\>

### Parameters

• **formmlSchema**: `Form`

• **preprocess?**: `Record`\<`PRIMITIVE`, `GenericSchema`\<`unknown`, `unknown`, `BaseIssue`\<`unknown`\>\>\>

### Returns

`v.StrictObjectSchema`\<`Record`\<`string`, `v.GenericSchema`\>, `string`\>

### Defined in

[validator/buildSchema.ts:49](https://github.com/formml/formml/blob/fed46848d8032d8aeab7f7fad75fbc02dc65656a/packages/core/src/validator/buildSchema.ts#L49)

## buildSchema(formmlSchema, preprocess)

> **buildSchema**(`formmlSchema`, `preprocess`?): `v.GenericSchema`

### Parameters

• **formmlSchema**: `Field` \| `Form`

• **preprocess?**: `Record`\<`PRIMITIVE`, `GenericSchema`\<`unknown`, `unknown`, `BaseIssue`\<`unknown`\>\>\>

### Returns

`v.GenericSchema`

### Defined in

[validator/buildSchema.ts:53](https://github.com/formml/formml/blob/fed46848d8032d8aeab7f7fad75fbc02dc65656a/packages/core/src/validator/buildSchema.ts#L53)
