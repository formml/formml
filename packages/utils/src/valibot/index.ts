export * from './schemas/required.js'
export * from './validations/isoDateTime.js'
export * from './validations/numerical.js'
export * from './validations/allowBlank.js'
export * from './validations/numeric-validations.js'
