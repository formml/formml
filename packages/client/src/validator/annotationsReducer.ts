import annotationsInterface from '@formml/dsl/interfaces/annotations.js'
import { GenericSchema, GenericValidation } from 'valibot'

import { assertNever } from '../utils/assertNever.js'
import { required } from './valibot/schemas/required.js'

type IAnnotations = typeof annotationsInterface

type Options<T extends readonly { name: string }[]> = {
  [K in T[number] as K['name']]: unknown
}

type IAnnotationAction = {
  [key in keyof IAnnotations]: {
    name: key
    options: Options<IAnnotations[key]>
  }
}[keyof IAnnotations]

type IValidationComponents = {
  pipeline: GenericValidation[]
  schema: GenericSchema
}

function annotationsReducer(
  components: IValidationComponents,
  action: IAnnotationAction,
): IValidationComponents {
  if (action.name === 'required') {
    return {
      ...components,
      schema: required(components.schema),
    }
  }
  return assertNever`Unknown annotation "${action.name}".`
}

export default annotationsReducer
