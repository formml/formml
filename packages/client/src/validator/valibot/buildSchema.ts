import { Field, PrimitiveType } from '@formml/dsl'
import { BigNumber } from 'bignumber.js'
import dayjs from 'dayjs'
import * as v from 'valibot'

import { PrimitiveTypeMapping } from '../../JsTypes.js'
import { assertNever } from '../../utils/assertNever.js'

const type = (
  formmlSchema: Field,
  options?: Partial<{
    notEmpty: boolean
  }>,
) => {
  if (formmlSchema.type === 'text') {
    if (options?.notEmpty) {
      return v.pipe(v.string(), v.nonEmpty())
    }
    return v.string()
  }
  if (formmlSchema.type === 'num') {
    return v.number()
  }
  if (formmlSchema.type === 'bool') {
    return v.boolean()
  }
  if (formmlSchema.type === 'datetime') {
    return v.custom<dayjs.Dayjs>(dayjs.isDayjs)
  }
  if (formmlSchema.type === 'decimal') {
    return v.instance(BigNumber)
  }
  return assertNever`Unsupported field schema: ${formmlSchema}`
}

const isRequired = (formmlSchema: Field) =>
  formmlSchema.annotations.some((a) => a.name === 'required')

export default function buildSchema<T extends PrimitiveType>(
  formmlSchema: Field<T>,
): v.GenericSchema<PrimitiveTypeMapping[T]>
export default function buildSchema(formmlSchema: Field) {
  if (isRequired(formmlSchema)) {
    return type(formmlSchema, { notEmpty: true })
  }
  return v.optional(type(formmlSchema))
}
