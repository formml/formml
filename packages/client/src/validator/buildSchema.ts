import { Annotation, Field, PrimitiveType, resolveArguments } from '@formml/dsl'
import annotationsInterface from '@formml/dsl/interfaces/annotations.js'
import { BigNumber } from 'bignumber.js'
import * as v from 'valibot'

import { PrimitiveTypeMapping } from '../JsTypes.js'
import { assertNever } from '../utils/assertNever.js'
import annotationsReducer, { IAnnotationAction } from './annotationsReducer.js'

const type = (formmlSchema: Field) => {
  if (formmlSchema.type === 'text') {
    return v.string()
  }
  if (formmlSchema.type === 'num') {
    return v.number()
  }
  if (formmlSchema.type === 'bool') {
    return v.boolean()
  }
  if (formmlSchema.type === 'datetime') {
    return v.date()
  }
  if (formmlSchema.type === 'decimal') {
    return v.instance(BigNumber)
  }
  return assertNever`Unsupported field schema: ${formmlSchema}`
}

function buildAction(annotation: Annotation): IAnnotationAction {
  // TODO: get info from real referred declaration and ensure type safety
  const annotationName = annotation.call.$refText as IAnnotationAction['name']
  return {
    name: annotationName,
    options: resolveArguments(
      annotation.args,
      annotationsInterface[annotationName],
    ),
  }
}

export default function buildSchema<T extends PrimitiveType>(
  formmlSchema: Field<T>,
): v.GenericSchema<PrimitiveTypeMapping[T]>
export default function buildSchema(formmlSchema: Field) {
  const baseSchema = v.optional(type(formmlSchema))
  const actions = formmlSchema.annotations.map(buildAction)
  return actions.reduce(annotationsReducer, baseSchema)
}
