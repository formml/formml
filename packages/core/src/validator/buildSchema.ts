import type { Annotation, Field, Form, PRIMITIVE } from '@formml/dsl'

import { isForm, utils } from '@formml/dsl'
import annotationsInterface from '@formml/dsl/interfaces/annotations.js'
import { assertNever } from '@formml/utils'
import { BigNumber } from 'bignumber.js'
import * as v from 'valibot'

import type { PrimitiveTypeMapping } from '../JsTypes.js'
import type { IAnnotationAction } from './annotationsReducer.js'

import annotationsReducer from './annotationsReducer.js'

const type = (formmlSchema: Field) => {
  if (formmlSchema.type === 'text') {
    return v.string()
  }
  if (formmlSchema.type === 'num') {
    return v.number()
  }
  if (formmlSchema.type === 'bool') {
    return v.boolean()
  }
  if (formmlSchema.type === 'datetime') {
    return v.date()
  }
  if (formmlSchema.type === 'decimal') {
    return v.instance(BigNumber)
  }
  return assertNever`Unsupported field schema: ${formmlSchema}`
}

function buildAction(annotation: Annotation): IAnnotationAction {
  // TODO: get info from real referred declaration and ensure type safety
  const annotationName = annotation.call.$refText as IAnnotationAction['name']
  return {
    name: annotationName,
    options: utils.resolveArguments(
      annotation.args,
      annotationsInterface[annotationName],
    ),
  }
}

export default function buildSchema<T extends PRIMITIVE>(
  formmlSchema: Field<T>,
): v.GenericSchema<PrimitiveTypeMapping[T]>
export default function buildSchema(
  formmlSchema: Form,
): v.StrictObjectSchema<Record<string, v.GenericSchema>, string>
export default function buildSchema(formmlSchema: Field | Form) {
  if (isForm(formmlSchema)) {
    const entries = formmlSchema.fields.map((field) => [
      field.name,
      buildSchema(field),
    ])
    return v.strictObject(Object.fromEntries(entries))
  }

  const baseSchema = v.optional(type(formmlSchema))
  const actions = formmlSchema.annotations.map(buildAction)
  return actions.reduce(annotationsReducer, baseSchema)
}
