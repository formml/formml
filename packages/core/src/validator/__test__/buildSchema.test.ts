import type { Field, Form, FormMLSchema } from '@formml/dsl'

import { BigNumber } from 'bignumber.js'
import dayjs from 'dayjs'
import * as v from 'valibot'

import annotationsReducer from '../annotationsReducer.js'
import buildSchema from '../buildSchema.js'

vi.mock('../annotationsReducer.js', async (importActual) => {
  const actual = await importActual<typeof import('../annotationsReducer.js')>()
  return {
    default: vi.fn(actual.default),
  }
})

describe('buildSchema', () => {
  describe('for field', () => {
    describe('integration', () => {
      describe.each(['text', 'num', 'bool', 'datetime', 'decimal'] as const)(
        'types - %s',
        (type) => {
          const fixtures = {
            bool: {
              expectedType: 'boolean',
              invalidInput: 'abc',
              validInput: true,
            },
            datetime: {
              expectedType: 'Date',
              invalidInput: dayjs('2024-01-01T00:00:00Z'),
              validInput: new Date('2024-01-01T00:00:00Z'),
            },
            decimal: {
              expectedType: 'BigNumber',
              invalidInput: 123.45,
              validInput: new BigNumber('123.45'),
            },
            num: {
              expectedType: 'number',
              invalidInput: 'abc',
              validInput: 123,
            },
            text: {
              expectedType: 'string',
              invalidInput: 123,
              validInput: 'abc',
            },
          }

          test(`should validate ${fixtures[type].expectedType} input if field type is ${type}`, () => {
            // Arrange
            const field: Field = {
              $container: {} as Form,
              $type: 'Field',
              annotations: [],
              name: 'field',
              type,
            }

            // Act
            const schema = buildSchema(field)
            const result = v.safeParse(schema, fixtures[type].validInput)

            // Assert
            expect(result.success).toBe(true)
          })

          test(`should invalidate non-${fixtures[type].expectedType} input if field type is ${type}`, () => {
            // Arrange
            const field: Field = {
              $container: {} as Form,
              $type: 'Field',
              annotations: [],
              name: 'field',
              type,
            }

            // Act
            const schema = buildSchema(field)
            const result = v.safeParse(schema, fixtures[type].invalidInput)

            // Assert
            expect(result.success).toBe(false)
            expect(result.issues).toMatchSnapshot()
          })
        },
      )

      test.each(['text', 'num', 'bool', 'datetime', 'decimal'] as const)(
        'should be optional by default',
        (type) => {
          // Arrange
          const field: Field = {
            $container: {} as Form,
            $type: 'Field',
            annotations: [],
            name: 'field',
            type,
          }

          // Act
          const schema = buildSchema(field)
          const result = v.safeParse(schema, undefined)

          // Assert
          expect(result.success).toBe(true)
        },
      )
    })

    describe('unit test', () => {
      test('should wrap base schema with annotations reducer', () => {
        // Arrange
        const field: Field = {
          $container: {} as Form,
          $type: 'Field',
          annotations: [
            {
              $container: {} as Field,
              $type: 'Annotation',
              args: [],
              call: {
                $refText: 'required',
              },
            },
          ],
          name: 'field',
          type: 'text',
        }
        const stubWrappedSchema = {} as v.GenericSchema
        vi.mocked(annotationsReducer).mockReturnValueOnce(stubWrappedSchema)

        // Act
        const schema = buildSchema(field)

        // Assert
        expect(annotationsReducer).toBeCalledWith(
          expect.anything(),
          {
            name: 'required',
            options: {},
          },
          expect.anything(),
          expect.anything(),
        )
        expect(schema).toBe(stubWrappedSchema)
      })
    })
  })

  describe('for form', () => {
    test('should wrap all fields with object schema', () => {
      // Arrange
      const form: Form = {
        $container: {} as FormMLSchema,
        $type: 'Form',
        fields: [
          {
            $container: {} as Form,
            $type: 'Field',
            annotations: [],
            name: 'textField',
            type: 'text',
          },
          {
            $container: {} as Form,
            $type: 'Field',
            annotations: [],
            name: 'numField',
            type: 'num',
          },
        ],
        name: 'ExampleForm',
      }

      // Act
      const schema = buildSchema(form)

      // Assert
      expect(schema).toEqual(
        expect.objectContaining({
          async: false,
          entries: {
            numField: expect.any(Object),
            textField: expect.any(Object),
          },
          kind: 'schema',
          type: 'strict_object',
        }),
      )
    })
  })
})
