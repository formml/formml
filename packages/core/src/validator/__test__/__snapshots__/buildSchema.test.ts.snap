// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildSchema > for field > integration > types - bool > should invalidate non-boolean input if field type is bool 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "boolean",
    "input": "abc",
    "issues": undefined,
    "kind": "schema",
    "lang": undefined,
    "message": "Invalid type: Expected boolean but received "abc"",
    "path": undefined,
    "received": ""abc"",
    "requirement": undefined,
    "type": "boolean",
  },
]
`;

exports[`buildSchema > for field > integration > types - datetime > should invalidate non-Date input if field type is datetime 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "Date",
    "input": "2024-01-01T00:00:00.000Z",
    "issues": undefined,
    "kind": "schema",
    "lang": undefined,
    "message": "Invalid type: Expected Date but received M",
    "path": undefined,
    "received": "M",
    "requirement": undefined,
    "type": "date",
  },
]
`;

exports[`buildSchema > for field > integration > types - decimal > should invalidate non-BigNumber input if field type is decimal 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "BigNumber",
    "input": 123.45,
    "issues": undefined,
    "kind": "schema",
    "lang": undefined,
    "message": "Invalid type: Expected BigNumber but received 123.45",
    "path": undefined,
    "received": "123.45",
    "requirement": undefined,
    "type": "instance",
  },
]
`;

exports[`buildSchema > for field > integration > types - num > should invalidate non-number input if field type is num 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "number",
    "input": "abc",
    "issues": undefined,
    "kind": "schema",
    "lang": undefined,
    "message": "Invalid type: Expected number but received "abc"",
    "path": undefined,
    "received": ""abc"",
    "requirement": undefined,
    "type": "number",
  },
]
`;

exports[`buildSchema > for field > integration > types - text > should invalidate non-string input if field type is text 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "string",
    "input": 123,
    "issues": undefined,
    "kind": "schema",
    "lang": undefined,
    "message": "Invalid type: Expected string but received 123",
    "path": undefined,
    "received": "123",
    "requirement": undefined,
    "type": "string",
  },
]
`;
